### PR TITLE
NAS-117563 / 22.12 / Add ZFS.userspace() for getting user quotas

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
I'll also open a middleware PR shortly.

File: libzfs.pyx [hunk 1]
```diff
@@ -1076,6 +1091,34 @@ cdef class ZFS(object):
         datasets = datasets or [p.name for p in self.pools]
         return ZFS._snapshots_serialized_impl(self.handle, datasets, props, holds, mounted, recursive, min_txg, max_txg)
 
+    @staticmethod
+    cdef int _userspace_cb(void *data, const char *domain, uint32_t rid, uint64_t space) nogil:
```
Comment (yocalebo) [inline libzfs.pyx]: Is there a reason to have function definition as `nogil` but inside it grabs the `gil` always?

Comment (yocalebo) [reply libzfs.pyx]: I think you can remove the `nogil` and then remove `with gil` statement so this is done while gil is held.

Comment (ghost) [reply libzfs.pyx]: I saw it done this way somewhere else, and I like that it's explicitly grabbing the gil inside this function as it is a callback invoked from a C context.

**Comment by anodos325** [comment]
Why not have as method under ZFSResource since we already have handle open there? Then we can eliminate path argument.

**Review by yocalebo — APPROVED** [review]

**Pull request merged**